File: core/src/com/bombbird/terminalcontrol/entities/sidstar/RandomSTAR.java
package com.bombbird.terminalcontrol.entities.sidstar;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.utils.Array;
import com.bombbird.terminalcontrol.entities.runways.Runway;
import com.bombbird.terminalcontrol.entities.airports.Airport;
import com.bombbird.terminalcontrol.entities.trafficmanager.DayNightManager;
import com.bombbird.terminalcontrol.utilities.saving.FileLoader;
import org.json.JSONObject;

import java.util.HashMap;

public class RandomSTAR {
    private static final HashMap<String, HashMap<String, Boolean>> noise = new HashMap<>();
    private static final HashMap<String, HashMap<String, Float>> time = new HashMap<>();

    /** Loads STAR noise info for the airport */
    public static void loadStarNoise(String icao) {
        noise.put(icao, FileLoader.loadNoise(icao, false));
    }

    /** Loads arrival entry timings */
    public static void loadEntryTiming(Airport airport) {
        HashMap<String, Float> stars = new HashMap<>();
        for (Star star: airport.getStars().values()) {
            stars.put(star.getName(), 0f);
        }
        time.put(airport.getIcao(), stars);
    }

    /** Loads arrival entry timings from save */
    public static void loadEntryTiming(Airport airport, JSONObject jsonObject) {
        for (String star: jsonObject.keySet()) {
            if (time.get(airport.getIcao()).containsKey(star)) {
                time.get(airport.getIcao()).put(star, (float) jsonObject.getDouble(star));
            }
        }
    }

    /** Updates timings */
    public static void update() {
        float dt = Gdx.graphics.getDeltaTime();
        for (String icao: time.keySet()) {
            for (String star: time.get(icao).keySet()) {
                time.get(icao).put(star, time.get(icao).get(star) - dt);
            }
        }
    }

    /** Gets a random STAR for the airport and runway */
    public static Star randomSTAR(Airport airport) {
        HashMap<String, Runway> rwys = airport.getLandingRunways();
        Array<Star> possibleStars = new Array<>();
        for (Star star: airport.getStars().values()) {
            boolean found = false;
            for (int i = 0; i < star.getRunways().size; i++) {
                if (rwys.containsKey(star.getRunways().get(i))) {
                    found = true;
                    break;
                }
            }
            if (found && checkNoise(airport, star.getName()) && time.get(airport.getIcao()).get(star.getName()) < 0) possibleStars.add(star);
        }

        if (possibleStars.size == 0) {
            Array<String> runways = new Array<>();
            for (String rwy: rwys.keySet()) {
                runways.add(rwy);
            }
            Gdx.app.log("Random STAR", "No STARs found to match criteria for " + airport.getIcao() + " " + runways.toString());
            throw new IllegalArgumentException("No STARs found to match criteria for " + airport.getIcao() + " " + runways.toString());
        } else {
            return possibleStars.random();
        }
    }

    /** Gets a list of all possible STARs that can be used with the current runway configuration */
    public static Array<String> getAllPossibleSTARnames(Airport airport) {
        Array<String> array = new Array<>();
        for (Star star: airport.getStars().values()) {
            boolean found = false;
            for (int i = 0; i < star.getRunways().size; i++) {
                if (airport.getLandingRunways().containsKey(star.getRunways().get(i))) {
                    found = true;
                    break;
                }
            }
            if (found && checkNoise(airport, star.getName())) array.add(star.getName() + " arrival");
        }

        return array;
    }

    /** Check whether a STAR is allowed to be used for the airport at the current time */
    private static boolean checkNoise(Airport airport, String star) {
        if (!noise.get(airport.getIcao()).containsKey(star)) return true; //Star can be used both during day, night
        return DayNightManager.checkNoiseAllowed(noise.get(airport.getIcao()).get(star));
    }

    /** Used to check if any STAR is available at airport, called before spawning new arrival */
    public static boolean starAvailable(Airport airport) {
        for (Star star: airport.getStars().values()) {
            if (time.get(airport.getIcao()).get(star.getName()) < 0 && checkNoise(airport, star.getName())) return true;
        }
        return false;
    }

    public static void starUsed(String icao, String star) {
        time.get(icao).put(star, 60f);
    }

    public static HashMap<String, HashMap<String, Float>> getTime() {
        return time;
    }
}

File: core/src/com/bombbird/terminalcontrol/entities/sidstar/RandomSTAR.kt
package com.bombbird.terminalcontrol.entities.sidstar

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.utils.Array
import com.bombbird.terminalcontrol.entities.airports.Airport
import com.bombbird.terminalcontrol.entities.runways.Runway
import com.bombbird.terminalcontrol.entities.trafficmanager.DayNightManager.checkNoiseAllowed
import com.bombbird.terminalcontrol.utilities.files.FileLoader
import org.json.JSONObject
import java.util.*

object RandomSTAR {
    private val noise = HashMap<String, HashMap<String, Boolean>>()
    val time = HashMap<String, HashMap<String, Float>>()

    /** Loads STAR noise info for the airport  */
    fun loadStarNoise(icao: String) {
        noise[icao] = FileLoader.loadNoise(icao, false)
    }

    /** Loads arrival entry timings  */
    fun loadEntryTiming(airport: Airport) {
        val stars = HashMap<String, Float>()
        for (star in airport.stars.values) {
            stars[star.name] = 0f
        }
        time[airport.icao] = stars
    }

    /** Loads arrival entry timings from save  */
    fun loadEntryTiming(airport: Airport, jsonObject: JSONObject) {
        for (star in jsonObject.keySet()) {
            if (time[airport.icao]?.containsKey(star) == true) {
                time[airport.icao]?.set(star, jsonObject.getDouble(star).toFloat())
            }
        }
    }

    /** Updates timings  */
    fun update() {
        val dt = Gdx.graphics.deltaTime
        for (icao in time.keys) {
            val airportStars = time[icao] ?: continue
            for (star in airportStars.keys) {
                val starTime = airportStars[star] ?: continue
                time[icao]?.set(star, starTime - dt)
            }
        }
    }

    /** Gets a random STAR for the airport and runway  */
    fun randomSTAR(airport: Airport): Star {
        val rwys: HashMap<String, Runway> = airport.landingRunways
        val possibleStars = Array<Star>()
        for (star in airport.stars.values) {
            var found = false
            for (i in 0 until star.runways.size) {
                if (rwys.containsKey(star.runways.get(i))) {
                    found = true
                    break
                }
            }
            if (found && checkNoise(airport, star.name) && (time[airport.icao]?.get(star.name) ?: 1f) < 0) possibleStars.add(star)
        }
        return if (possibleStars.size == 0) {
            val runways = Array<String>()
            for (rwy in rwys.keys) {
                runways.add(rwy)
            }
            Gdx.app.log("Random STAR", "No STARs found to match criteria for " + airport.icao + " " + runways.toString())
            throw IllegalArgumentException("No STARs found to match criteria for " + airport.icao + " " + runways.toString())
        } else {
            possibleStars.random()
        }
    }

    /** Gets a list of all possible STARs that can be used with the current runway configuration  */
    fun getAllPossibleSTARnames(airport: Airport): Array<String> {
        val array = Array<String>()
        for (star in airport.stars.values) {
            var found = false
            for (i in 0 until star.runways.size) {
                if (airport.landingRunways.containsKey(star.runways.get(i))) {
                    found = true
                    break
                }
            }
            if (found && checkNoise(airport, star.name)) array.add(star.name + " arrival")
        }
        return array
    }

    /** Check whether a STAR is allowed to be used for the airport at the current time  */
    private fun checkNoise(airport: Airport, star: String): Boolean {
        return noise[airport.icao]?.get(star)?.let { checkNoiseAllowed(it) } ?: true //Star can be used both during day, night if not in hashMap
    }

    /** Used to check if any STAR is available at airport, called before spawning new arrival  */
    fun starAvailable(airport: Airport): Boolean {
        for (star in airport.stars.values) {
            if ((time[airport.icao]?.get(star.name) ?: 1f) < 0 && checkNoise(airport, star.name)) return true
        }
        return false
    }

    /** Resets STAR timer to 60 seconds after arrival spawns there */
    fun starUsed(icao: String, star: String) {
        time[icao]?.set(star, 60f)
    }
}